### PR TITLE
Try to simplify FloorDiv axioms implications when needed during evaluations.

### PR DESCRIPTION
**Pull request by laithsakka**
Summary:
This very much the same solution proposed by bobrenjc93 except that it restrict it to expressions and axioms that have FloorDiv, since those are the only ones that could have became CleanDiv. and the only one that can changes as shape env changes. 

This also does not break torchrec benchmarks, it might be worth it to know why the generalization of this does break the torchrec benchmarks, but we could just be hitting another bug or NYI situation.

ovearhead?
None on
```
buck2 run fbcode//mode/opt fbcode//torchrec/distributed/tests:pt2_compile_benchmark -- --num-features=1000
```

Differential Revision: D66307433

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -5389,22 +5389,35 @@ def _maybe_evaluate_static(
...
         expr = canonicalize_bool_expr(expr)
 
+        def try_simplify_floor_div(axioms):
+            for k, v in axioms.items():
+                # A FloorDiv in implications could have became CleanDiv at this point, due to new facts
+                # to the sahpeEnv. This handles such issue but its not ideal. This is the only expression
+                # simplification that depends on the global state of shape env.
+                # TODO try to get rid of CleanDiv since its breaks the invariant thats simplifications of sympy
+                # expressions only depend on the expression itself.
+                if k.has(FloorDiv) and k.free_symbols.issubset(expr.free_symbols):
```
Comment (laithsakka) [inline torch/fx/experimental/symbolic_shapes.py:5406]: check if its more efificient to remove 
and k.free_symbols.issubset(expr.free_symbols):

**Comment by laithsakka** [comment]
removing  k.free_symbols.issubset(expr.free_symbols) remove the regression on torch rec. update the PR

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -5389,22 +5389,35 @@ def _maybe_evaluate_static(
...
 
         if compute_hint:
             expr = expr.xreplace(self.var_to_val).xreplace(self.unbacked_var_to_val)
 
         expr = canonicalize_bool_expr(expr)
 
+        def try_simplify_floor_div(axioms):
+            for k, v in axioms.items():
+                # A FloorDiv in implications could have became CleanDiv at this point, due to new facts
+                # to the sahpeEnv. This handles such issue but its not ideal. This is the only expression
```
Comment (bobrenjc93) [inline torch/fx/experimental/symbolic_shapes.py:5402]: s/sahpeEnv/shapeEnv

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -5389,22 +5389,35 @@ def _maybe_evaluate_static(
...
+                # simplification that depends on the global state of shape env.
+                # TODO try to get rid of CleanDiv since its breaks the invariant thats simplifications of sympy
```
Comment (bobrenjc93) [inline torch/fx/experimental/symbolic_shapes.py:5404]: s/since its breaks/since it breaks

**Comment by bobrenjc93** [comment]
Can you also add a test?

**Comment by laithsakka** [comment]
> Reference in ne

good ideal, let me try its a food practice to figure out how to write unit test for those things

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -5344,22 +5344,37 @@ def _maybe_evaluate_static(
...
+                # TODO try to get rid of CleanDiv since it breaks the invariant thats simplifications of sympy
+                # expressions only depend on the expression itself.
+                if k.has(FloorDiv):
+                    new_items.update({self.simplify(k): v})
+            axioms.update(new_items)
+
         # Pattern matching
         if axioms is None:
+            if expr.has(CleanDiv):
+                try_simplify_floor_div(self.axioms)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:5369]: Why don't you just unconditionally run this after the if statement

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -5344,22 +5344,37 @@ def _maybe_evaluate_static(
...
+            new_items = {}
+            for k, v in axioms.items():
+                # A FloorDiv in implications could have became CleanDiv at this point, due to new facts
+                # to the shapeEnv. This handles such issue but its not ideal. This is the only expression
+                # simplification that depends on the global state of shape env.
+                # TODO try to get rid of CleanDiv since it breaks the invariant thats simplifications of sympy
+                # expressions only depend on the expression itself.
+                if k.has(FloorDiv):
+                    new_items.update({self.simplify(k): v})
+            axioms.update(new_items)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:5364]: I think a slight optimization is that you only have to do these updates when a new divisibility constraint is added; e.g., if I keep doing CleanDiv over and over again, I don't have to keep trying to simplify here (I think it's relatively simple to make a synthetic test that would keep hitting `try_simplify_floor_div` here).

**Review by ezyang — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py (hunk 5 above)
Comment (laithsakka) [reply torch/fx/experimental/symbolic_shapes.py:5364]: good idea let me do that

File: torch/fx/experimental/symbolic_shapes.py (hunk 4 above)
Comment (laithsakka) [reply torch/fx/experimental/symbolic_shapes.py:5369]: I am not sure if its faster to  expr.has(CleanDiv): or iterate on the axioms.  for each expression
but if we do change were we only run that when a new divisibility constraint is added. then probably always running it is ok

Comment (laithsakka) [reply torch/fx/experimental/symbolic_shapes.py:5369]: Actually removing expr.has(CleanDiv) makes it very slow!

File: torch/fx/experimental/symbolic_shapes.py (hunk 5 above)
Comment (laithsakka) [reply torch/fx/experimental/symbolic_shapes.py:5364]: here is a situation were that can happen and be expensive, 
```
        from sympy import Eq

        test1 = Eq(FloorDiv(a.node.expr, 3), 1)
        test2 = Eq(CleanDiv(a.node.expr, 3), 1)

        self.assertTrue(shape_env.evaluate_expr(test1))
        self.assertEqual(shape_env._maybe_evaluate_static(test2), None)

        # After this FloorDiv(a, 3) is simplified to CleanDiv(a, 3)
        shape_env.defer_runtime_assert(Eq(Mod(a, 3), 0), " test")
        self.assertEqual(test2, shape_env.simplify(test1))
        #25692715671
        for i in range(1, 1000):
            # add unrelated change to shape env.
            a = shape_env.create_unbacked_symint()
            shape_env.defer_runtime_assert((a // 3 == 1).node.expr, " test")

        from torch._C import _instruction_counter
    
        start = _instruction_counter.start()

        print(start)
        for i in range(1, 1000):
            # add unrelated change to shape env.
            a = shape_env.create_unbacked_symint()
            shape_env.defer_runtime_assert((a== 1).node.expr, " test")
            a = shape_env.create_unbacked_symint()
            shape_env._maybe_evaluate_static(CleanDiv(a.node.expr, i))

        end = _instruction_counter.end(start)
        print(end)
```

**Comment by laithsakka** [comment]
address comments and rebase

**Pull request closed without merging**